File: src/test/java/im/tox/gui/InvokeLaterToxEventListener.java
package im.tox.gui;

import im.tox.tox4j.annotations.NotNull;
import im.tox.tox4j.core.callbacks.ToxEventListener;
import im.tox.tox4j.core.enums.ToxConnection;
import im.tox.tox4j.core.enums.ToxFileControl;
import im.tox.tox4j.core.enums.ToxMessageType;
import im.tox.tox4j.core.enums.ToxUserStatus;

import javax.swing.*;

@SuppressWarnings("checkstyle:linelength")
public final class InvokeLaterToxEventListener implements ToxEventListener {

  @Override
  public void selfConnectionStatus(@NotNull final ToxConnection connectionStatus) {
    SwingUtilities.invokeLater(new Runnable() {
      @Override
      public void run() {
        underlying.selfConnectionStatus(connectionStatus);
      }
    });
  }

  @Override
  public void fileRecvControl(final int friendNumber, final int fileNumber, @NotNull final ToxFileControl control) {
    SwingUtilities.invokeLater(new Runnable() {
      @Override
      public void run() {
        underlying.fileRecvControl(friendNumber, fileNumber, control);
      }
    });
  }

  @Override
  public void fileRecv(final int friendNumber, final int fileNumber, final int kind, final long fileSize, @NotNull final byte[] filename) {
    SwingUtilities.invokeLater(new Runnable() {
      @Override
      public void run() {
        underlying.fileRecv(friendNumber, fileNumber, kind, fileSize, filename);
      }
    });
  }

  @Override
  public void fileRecvChunk(final int friendNumber, final int fileNumber, final long position, @NotNull final byte[] data) {
    SwingUtilities.invokeLater(new Runnable() {
      @Override
      public void run() {
        underlying.fileRecvChunk(friendNumber, fileNumber, position, data);
      }
    });
  }

  @Override
  public void fileChunkRequest(final int friendNumber, final int fileNumber, final long position, final int length) {
    SwingUtilities.invokeLater(new Runnable() {
      @Override
      public void run() {
        underlying.fileChunkRequest(friendNumber, fileNumber, position, length);
      }
    });
  }

  @Override
  public void friendConnectionStatus(final int friendNumber, @NotNull final ToxConnection connectionStatus) {
    SwingUtilities.invokeLater(new Runnable() {
      @Override
      public void run() {
        underlying.friendConnectionStatus(friendNumber, connectionStatus);
      }
    });
  }

  @Override
  public void friendLosslessPacket(final int friendNumber, @NotNull final byte[] data) {
    SwingUtilities.invokeLater(new Runnable() {
      @Override
      public void run() {
        underlying.friendLosslessPacket(friendNumber, data);
      }
    });
  }

  @Override
  public void friendLossyPacket(final int friendNumber, @NotNull final byte[] data) {
    SwingUtilities.invokeLater(new Runnable() {
      @Override
      public void run() {
        underlying.friendLossyPacket(friendNumber, data);
      }
    });
  }

  @Override
  public void friendMessage(final int friendNumber, @NotNull final ToxMessageType type, final int timeDelta, @NotNull final byte[] message) {
    SwingUtilities.invokeLater(new Runnable() {
      @Override
      public void run() {
        underlying.friendMessage(friendNumber, type, timeDelta, message);
      }
    });
  }

  @Override
  public void friendName(final int friendNumber, @NotNull final byte[] name) {
    SwingUtilities.invokeLater(new Runnable() {
      @Override
      public void run() {
        underlying.friendName(friendNumber, name);
      }
    });
  }

  @Override
  public void friendRequest(@NotNull final byte[] publicKey, final int timeDelta, @NotNull final byte[] message) {
    SwingUtilities.invokeLater(new Runnable() {
      @Override
      public void run() {
        underlying.friendRequest(publicKey, timeDelta, message);
      }
    });
  }

  @Override
  public void friendStatus(final int friendNumber, @NotNull final ToxUserStatus status) {
    SwingUtilities.invokeLater(new Runnable() {
      @Override
      public void run() {
        underlying.friendStatus(friendNumber, status);
      }
    });
  }

  @Override
  public void friendStatusMessage(final int friendNumber, @NotNull final byte[] message) {
    SwingUtilities.invokeLater(new Runnable() {
      @Override
      public void run() {
        underlying.friendStatusMessage(friendNumber, message);
      }
    });
  }

  @Override
  public void friendTyping(final int friendNumber, final boolean isTyping) {
    SwingUtilities.invokeLater(new Runnable() {
      @Override
      public void run() {
        underlying.friendTyping(friendNumber, isTyping);
      }
    });
  }

  @Override
  public void friendReadReceipt(final int friendNumber, final int messageId) {
    SwingUtilities.invokeLater(new Runnable() {
      @Override
      public void run() {
        underlying.friendReadReceipt(friendNumber, messageId);
      }
    });
  }


  private final ToxEventListener underlying;

  public InvokeLaterToxEventListener(ToxEventListener underlying) {
    this.underlying = underlying;
  }

}

File: projects/tox4j/src/test/java/im/tox/gui/util/InvokeLaterToxEventListener.scala
package im.tox.gui.util

import javax.swing._

import im.tox.tox4j.annotations.NotNull
import im.tox.tox4j.core.callbacks.ToxEventListener
import im.tox.tox4j.core.enums.{ToxConnection, ToxFileControl, ToxMessageType, ToxUserStatus}

// scalastyle:off line.size.limit
final class InvokeLaterToxEventListener[ToxCoreState](underlying: ToxEventListener[ToxCoreState]) extends ToxEventListener[ToxCoreState] {

  private def invokeLater(callback: ToxCoreState => ToxCoreState)(state: ToxCoreState): ToxCoreState = {
    SwingUtilities.invokeLater(new Runnable {
      override def run(): Unit = {
        callback(state)
      }
    })
    state
  }

  override def selfConnectionStatus(@NotNull connectionStatus: ToxConnection)(state: ToxCoreState): ToxCoreState = {
    invokeLater(underlying.selfConnectionStatus(connectionStatus))(state)
  }

  override def fileRecvControl(friendNumber: Int, fileNumber: Int, @NotNull control: ToxFileControl)(state: ToxCoreState): ToxCoreState = {
    invokeLater(underlying.fileRecvControl(friendNumber, fileNumber, control))(state)
  }

  override def fileRecv(friendNumber: Int, fileNumber: Int, kind: Int, fileSize: Long, @NotNull filename: Array[Byte])(state: ToxCoreState): ToxCoreState = {
    invokeLater(underlying.fileRecv(friendNumber, fileNumber, kind, fileSize, filename))(state)
  }

  override def fileRecvChunk(friendNumber: Int, fileNumber: Int, position: Long, @NotNull data: Array[Byte])(state: ToxCoreState): ToxCoreState = {
    invokeLater(underlying.fileRecvChunk(friendNumber, fileNumber, position, data))(state)
  }

  override def fileChunkRequest(friendNumber: Int, fileNumber: Int, position: Long, length: Int)(state: ToxCoreState): ToxCoreState = {
    invokeLater(underlying.fileChunkRequest(friendNumber, fileNumber, position, length))(state)
  }

  override def friendConnectionStatus(friendNumber: Int, @NotNull connectionStatus: ToxConnection)(state: ToxCoreState): ToxCoreState = {
    invokeLater(underlying.friendConnectionStatus(friendNumber, connectionStatus))(state)
  }

  override def friendLosslessPacket(friendNumber: Int, @NotNull data: Array[Byte])(state: ToxCoreState): ToxCoreState = {
    invokeLater(underlying.friendLosslessPacket(friendNumber, data))(state)
  }

  override def friendLossyPacket(friendNumber: Int, @NotNull data: Array[Byte])(state: ToxCoreState): ToxCoreState = {
    invokeLater(underlying.friendLossyPacket(friendNumber, data))(state)
  }

  override def friendMessage(friendNumber: Int, @NotNull `type`: ToxMessageType, timeDelta: Int, @NotNull message: Array[Byte])(state: ToxCoreState): ToxCoreState = {
    invokeLater(underlying.friendMessage(friendNumber, `type`, timeDelta, message))(state)
  }

  override def friendName(friendNumber: Int, @NotNull name: Array[Byte])(state: ToxCoreState): ToxCoreState = {
    invokeLater(underlying.friendName(friendNumber, name))(state)
  }

  override def friendRequest(@NotNull publicKey: Array[Byte], timeDelta: Int, @NotNull message: Array[Byte])(state: ToxCoreState): ToxCoreState = {
    invokeLater(underlying.friendRequest(publicKey, timeDelta, message))(state)
  }

  override def friendStatus(friendNumber: Int, @NotNull status: ToxUserStatus)(state: ToxCoreState): ToxCoreState = {
    invokeLater(underlying.friendStatus(friendNumber, status))(state)
  }

  override def friendStatusMessage(friendNumber: Int, @NotNull message: Array[Byte])(state: ToxCoreState): ToxCoreState = {
    invokeLater(underlying.friendStatusMessage(friendNumber, message))(state)
  }

  override def friendTyping(friendNumber: Int, isTyping: Boolean)(state: ToxCoreState): ToxCoreState = {
    invokeLater(underlying.friendTyping(friendNumber, isTyping))(state)
  }

  override def friendReadReceipt(friendNumber: Int, messageId: Int)(state: ToxCoreState): ToxCoreState = {
    invokeLater(underlying.friendReadReceipt(friendNumber, messageId))(state)
  }

}
